### PR TITLE
Updates flesch reading.

### PR DESCRIPTION
**Pull request by terw-dan**
This makes the fleschreading more accurate by running the syllable count per word, in stead of per text. This makes the use of regexes possible that match on the end of a string and improves are syllable count for words. 

Removes the use of cleantext, since we do a per word check. 

Refactored the count syllables function to use an object for the exceptions, so we can run multiple counters, in stead of only 1 for add and 1 for subtract.

Improves the exclusion regex for english.

Currently the objects we use are named languagesyllableregex, we need a better name for this. 

For testing: The most important thing is that the syllables in words get counted correctly. The easiest way to check this is to add a `console.log( word, syllableCount )` between rule 74 and 75, so it will log the word and the number of syllables found.

### CODE REVIEW

File: js/researches/calculateFleschReading.js [hunk 1]
```diff
@@ -22,21 +20,19 @@ module.exports = function( paper ) {
 		return 0;
 	}
 
-	var sentenceCount = countSentences( text );
+	var numberOfSentences = countSentences ( text );
```
Comment (boblinthorst) [inline js/researches/calculateFleschReading.js]: Remove whitespace between countSentence and the argument.

File: js/stringProcessing/countSyllables.js [hunk 2]
```diff
@@ -1,133 +1,99 @@
...
-
-		// split on consonants
-		splitWord = array[ i ].split( /[^aeiouy]/g );
-
-		// if the string isn't empty, a consonant was found, up the counter
-		for ( j = 0; j < splitWord.length; j++ ) {
-			if ( splitWord[ j ] !== "" ) {
-				count++;
-			}
+var countExclusions = function( word ) {
```
Comment (boblinthorst) [inline js/stringProcessing/countSyllables.js]: Change name to CountSyllablesInExclusionWord or something like it. Now it sounds like it returns the amount of exclusion words in the passed argument.

Comment (atimmer) [inline js/stringProcessing/countSyllables.js]: I would vote for `countExcludedWords`. Words that are excluded from the method using vowels.

Comment (IreneStr) [inline js/stringProcessing/countSyllables.js]: But you are not counting the words. You're counting the syllables in these words.

Comment (atimmer) [inline js/stringProcessing/countSyllables.js]: `countSyllablesInExcludedWords`? 😅

File: js/values/LanguageSyllableRegex.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	}
+};
+
+/**
+ * Returns the stored regular expression.
+ *
+ * @returns {RegExp} The stored regular expression.
+ */
+LanguageSyllableRegex.prototype.getRegex = function() {
+	return this._regex;
```
Comment (atimmer) [inline js/values/LanguageSyllableRegex.js]: This is not covered by the specs.

File: js/values/LanguageSyllableRegex.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+ *
+ * @param {String} word The word to match for syllable exclusions.
+ * @returns {number} The amount of syllables found.
+ */
+LanguageSyllableRegex.prototype.countSyllables = function( word ) {
+	if( this._hasRegex ) {
+		var match = word.match( this._regex ) || [];
+		return match.length * this._multiplier;
+	}
+	return 0;
```
Comment (atimmer) [inline js/values/LanguageSyllableRegex.js]: This line is not covered by the specs.

File: js/values/syllableCountMethod.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
+var LanguageSyllableRegex = require ( "./syllableCountStep.js" );
```
Comment (atimmer) [inline js/values/syllableCountMethod.js]: Please be diligent and make sure you rename all occurences.

File: js/values/syllableCountMethod.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+var isUndefined = require( "lodash/isUndefined" );
+var forEach = require ( "lodash/forEach" );
+
+/**
+ * Creates a Language Syllable Regex Master
+ *
+ * @param {object} config The config object containing an array with syllable exclusions
+ * @constructor
+ */
+var syllableCountIterator = function( config ) {
```
Comment (atimmer) [inline js/values/syllableCountMethod.js]: I love `Iterator` as name for this but I prefer the filename to reflect this too.

File: js/stringProcessing/countSyllables.js [hunk 7]
```diff
@@ -1,133 +1,99 @@
...
-var exclusionWords = syllableArray().exclusionWords;
-var exclusionWordsRegexes = map( exclusionWords, function( exclusionWord ) {
-	return {
-		regex: new RegExp( exclusionWord.word, "ig" ),
-		syllables: exclusionWord.syllables
-	};
-} );
-var addSyllablesRegex = arrayToRegex( syllableArray().addSyllables, true );
-var subtractSyllablesRegex = arrayToRegex( syllableArray().subtractSyllables, true );
+var LanguageSyllableRegexMaster = require( "../values/syllableCountMethod.js" );
```
Comment (atimmer) [inline js/stringProcessing/countSyllables.js]: Please be diligent and make sure you rename all occurrences.

File: js/stringProcessing/countSyllables.js [hunk 8]
```diff
@@ -1,133 +1,99 @@
...
+
+var languageSyllableRegexMaster = new LanguageSyllableRegexMaster( syllableMatchers() );
```
Comment (atimmer) [inline js/stringProcessing/countSyllables.js]: Please be diligent and make sure you rename all occurrences.

File: js/values/syllableCountMethod.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+	this.availableLanguageSyllableRegexes = [];
```
Comment (atimmer) [inline js/values/syllableCountMethod.js]: Why not call this `countSteps`?

File: js/values/syllableCountMethod.js [hunk 10]
```diff
@@ -0,0 +1,53 @@
+var LanguageSyllableRegex = require ( "./syllableCountStep.js" );
+
+var isUndefined = require( "lodash/isUndefined" );
+var forEach = require ( "lodash/forEach" );
+
+/**
+ * Creates a Language Syllable Regex Master
```
Comment (atimmer) [inline js/values/syllableCountMethod.js]: Please be diligent and make sure you rename all occurrences.

File: js/values/syllableCountStep.js [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+ *
+ * @param {String} word The word to match for syllable exclusions.
+ * @returns {number} The amount of syllables found.
+ */
+syllableCountStep.prototype.countSyllables = function( word ) {
+	if( this._hasRegex ) {
+		var match = word.match( this._regex ) || [];
+		return match.length * this._multiplier;
+	}
+	return 0;
```
Comment (atimmer) [inline js/values/syllableCountStep.js]: This line is not covered by the specs.

File: spec/stringProcessing/countSyllableSpec.js [hunk 12]
```diff
@@ -1,15 +1,25 @@
...
+		expect( countSyllableFunction("A test based on exclusion words for syllable count") ).toBe( 13 );
 	});
 });
+
+describe( "counting syllables", function() {
+	it( "returns the number of syllables in an exclusion word", function(){
+		expect( countSyllableFunction( "shoreline") ).toBe( 2 );
+		expect( countSyllableFunction( "business" ) ).toBe( 2 );
+	});
+	it( "returns the number of syllables in normal word", function(){
```
Comment (atimmer) [inline spec/stringProcessing/countSyllableSpec.js]: We shouldn't leave empty specs hanging around.

File: spec/stringProcessing/removeSentenceTerminitorsSpec.js [hunk 13]
```diff
@@ -3,15 +3,15 @@ var removeSentenceTerminators = require( "../../js/stringProcessing/removeSenten
 describe( "a test removing sentence terminators from a string", function() {
```
Comment (atimmer) [inline spec/stringProcessing/removeSentenceTerminitorsSpec.js]: Please rename this file without typos

File: js/stringProcessing/removeSentenceTerminators.js [hunk 14]
```diff
@@ -9,5 +9,5 @@ var sentenceTerminators = /[.?!:;,]/g;
  * @returns {String} The sanitized text.
  */
 module.exports = function( text ) {
-	return text.replace( sentenceTerminators, "" );
+	return text.replace( sentenceTerminators, " " );
```
Comment (atimmer) [inline js/stringProcessing/removeSentenceTerminators.js]: Please revert this and solve the problem that has been solved with this function in `getWords` in another way.

File: js/stringProcessing/countSyllables.js [hunk 15]
```diff
@@ -1,133 +1,99 @@
...
+ * @returns {number} The number of syllables found in the given word.
  */
-var removeExclusionWords = function( text ) {
-	forEach( exclusionWordsRegexes, function( exclusionWordRegex ) {
-		text = text.replace( exclusionWordRegex.regex, "" );
-	} );
-
-	return text;
+var countVowelExclusions = function( word ) {
+	return languageSyllableRegexMaster.countSyllables( word );
```
Comment (atimmer) [inline js/stringProcessing/countSyllables.js]: Please be diligent and make sure you rename all occurrences.

File: spec/values/languageSyllableRegexMasterSpec.js [hunk 16]
```diff
@@ -0,0 +1,29 @@
+var languageSyllableRegexMaster = require(  "../../js/values/syllableCountMethod.js" );
```
Comment (atimmer) [inline spec/values/languageSyllableRegexMasterSpec.js]: Please be diligent and make sure you rename all occurrences.

File: spec/values/languageSyllableRegexMasterSpec.js [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+
+describe( "creating a language syllable regex master", function() {
+	it( "returns an empty language syllable regex master", function() {
+		var mockMaster = new languageSyllableRegexMaster();
```
Comment (atimmer) [inline spec/values/languageSyllableRegexMasterSpec.js]: Please be diligent and make sure you rename all occurrences.

File: spec/values/languageSyllableRegexSpec.js [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+		var languageSyllableRegex = new LanguageSyllableRegex();
+		expect( languageSyllableRegex.hasRegex() ).toBe( false );
+	} );
+
+	it( "creates an language syllable regex with a +1 multiplier", function() {
+		var mockSyllables = {
+				syllables: [ "a" ],
+				multiplier: +1
+		};
+		var languageSyllableRegex = new LanguageSyllableRegex( mockSyllables );
```
Comment (atimmer) [inline spec/values/languageSyllableRegexSpec.js]: Please be diligent and make sure you rename all occurrences.

File: spec/values/languageSyllableRegexSpec.js [hunk 19]
```diff
@@ -0,0 +1,31 @@
+var LanguageSyllableRegex = require( "../../js/values/syllableCountStep.js" );
```
Comment (atimmer) [inline spec/values/languageSyllableRegexSpec.js]: Please be diligent and make sure you rename all occurrences.

**Comment by atimmer** [comment]
CR Done 👍🏻

**Pull request merged**